packages/apollo-server-core/src/utils/dispatcher.ts

```diff
@@ -1,19 +1,34 @@
-import { GraphQLRequestListener } from "apollo-server-plugin-base";
+import { AnyFunction, AnyFunctionMap } from "apollo-server-types";
 
-type AnyFunction = (...args: any[]) => any;
 type Args<F> = F extends (...args: infer A) => any ? A : never;
 type AsFunction<F> = F extends AnyFunction ? F : never;
 type UnwrapPromise<T> = T extends Promise<infer U> ? U : T;
 
 type DidEndHook<TArgs extends any[]> = (...args: TArgs) => void;
 
-export class Dispatcher<T extends GraphQLRequestListener> {
+export class Dispatcher<T extends AnyFunctionMap> {
   constructor(protected targets: T[]) {}
 
+  private callTargets<TMethodName extends keyof T>(
+    targets: T[],
+    methodName: TMethodName,
+    ...args: Args<T[TMethodName]>
+  ) {
```

Comment:
Would it be possible to write the return type here?

Reply:
Hmm.  Good question.  The inattention to adding a type here is certainly an oversight on my part, but I just refactored this into this shape utilizing this helper.  Looking at the previous state of the `dispatcher`, it would seem that the return values from target invocations is still as `any[]` typed as this is now.

Unless I'm mistaken, that's certainly not ideal, but I just spent a fair bit trying to make this feel more correct and I couldn't convince myself of a solution that was more correct.  If you have any suggestions, they'd be very welcome, but things would seem to get a little bit blurry around whether the return types are "maybe Promise"s"? 🤔

In practice, outside of the dispatcher, the type is correct by the invokers of the `public` dispatch methods within the `requestPipeline`. 

I'm happy to put things back to how they were to keep it out of the scope of this PR, I'd rather not have this be contingent on existing decisions if it's not making anything less type-safe.

Reply:
Oh, I didn't realize that it was `any[]`; I thought there was an easy thing involving `ReturnType<AsFunction<>>` like the others had. But I suppose those only work because of nested `any[]`!  nbd (and I think having the helper is still helpful).

I suppose we can at least reduce the scope of `any`ness by this?

```
diff --git a/packages/apollo-server-core/src/utils/dispatcher.ts b/packages/apollo-server-core/src/utils/dispatcher.ts
index 0ea76170..c497ee3d 100644
--- a/packages/apollo-server-core/src/utils/dispatcher.ts
+++ b/packages/apollo-server-core/src/utils/dispatcher.ts
@@ -13,7 +13,7 @@ export class Dispatcher<T extends AnyFunctionMap> {
     targets: T[],
     methodName: TMethodName,
     ...args: Args<T[TMethodName]>
-  ) {
+  ): ReturnType<AsFunction<T[TMethodName]>>[] {
     return targets.map(target => {
       const method = target[methodName];
       if (method && typeof method === 'function') {
@@ -27,7 +27,7 @@ export class Dispatcher<T extends AnyFunctionMap> {
     ...args: Args<T[TMethodName]>
   ): Promise<UnwrapPromise<ReturnType<AsFunction<T[TMethodName]>>>[]> {
     return await Promise.all(
-      this.callTargets(this.targets, methodName, ...args),
+      this.callTargets(this.targets, methodName, ...args) as any[],
     );
   }
```

nbd either way though